File: Codes/Matlab/CSP/24940822CSP(m_function)/loadov.m
function [signal3d, signal_label, Fs] = loadov(path)
% 输出格式
% signal3d : samples×channals×trials
% signal_label : trials
load(path);
Fs = samplingFreq ; % 采样 500Hz
data_x = samples;  % 数据
data_x_time = sampleTime;
data_x_time = data_x_time(:, 1);

% ==滑窗前带通滤波==
filt_n = 4; % 带通滤波器阶数
filter_low = 0.1; % 带通滤波范围
filter_high = 32;
data_x = BPfilter(data_x, Fs, filt_n, filter_low, filter_high);
% %===CAR===(某一采样点全部通道减均值)
channel_num = size(data_x, 2);
channel_sum = zeros(size(data_x, 1), 1);
for i = 1:channel_num
    channel_sum = channel_sum + data_x(:, i);
end
channel_mean = channel_sum / channel_num;
for i = 1:channel_num
    data_x(:, i) = data_x(:, i) - channel_mean;
end

% 确定 769/770 cue 出现位置
label = [];
label_time= [];
trialsize = 0;
for i = 1:size(stims,1)
    if stims(i, 2) == 769 || stims(i, 2) == 770
        trialsize = trialsize + 1;
        label = [label stims(i, 2)-769];
        label_time = [label_time stims(i, 1)];
    end
end
pos = zeros(1,length(data_x_time));
k = 1;
for i = 1:length(label_time)
    for j = 1:length(data_x_time)
        if data_x_time(j) > label_time(i)
            pos(k) = j;
            k = k + 1;
            break
        end
    end
end

width = 2500;
step = 10;
window_size = 500;
window_num = int32((width - window_size) / step + 1);
j = 1;

signal_label = zeros(1, trialsize*window_num);
channal_num = size(data_x, 2);
signal3d = zeros([window_size, channal_num, trialsize*window_num]);
for k = 1:length(label)
    for i = 1:window_num
        signal_label(j) = label(k);
        start = pos(k) + (i-1) * step;
        signal3d(:, :, j) = data_x(start:start + window_size - 1, :);
        j = j + 1;
    end
end
end

File: Codes/PythonProjects/loadData/loadov.py
# -*- coding: utf-8 -*-
import scipy.io as sio
import numpy as np
import matplotlib.pyplot as plt
from sklearn import preprocessing

def loadov(path):
    """
       读NE生成的ov转mat文件，生成滑窗后数据

       输入参数
       ----------
       filepath: npz文件路径

       返回值
       ----------
       signal3d: T×N×L ndarray
                 T: 采样点数  N: 通道数  L: 训练数据trial总数
       signal_label: shape (n_samples,)
                L 个trial对应的标签


       """
    data = sio.loadmat(path)
    Fs = data['samplingFreq']  # 采样 500Hz
    data_x = data['samples']  # 数据
    data_x_time = data['sampleTime']
    data_x_time = data_x_time[:, 0].tolist()
    stims = data['stims']  # 标签[time; ID; duration] (10个769,10个770)

    # 确定 769/770 cue 出现位置
    label = []
    label_time= []
    trialsize = 0
    for i in range(stims.shape[0]):
        if stims[i, 1] == 769 or stims[i, 1] == 770:
            trialsize = trialsize + 1
            label.append(stims[i, 1]-769)
            label_time.append(stims[i, 0])
    pos = []
    for i in range(len(label_time)):
        for j in range(len(data_x_time)):
            if data_x_time[j] > label_time[i]:
                pos.append(j)
                break


    # 滑窗
    width = 2500  # 单个trial 的总长
    step = 50  # 步长
    window_size = 500  # 窗的大小
    delay = 500  # cue 出现后延迟
    window_num = int((width - delay - window_size) / step + 1)  # 每个trial 滑多少个窗
    j = 0
    signal_label = np.zeros(trialsize * window_num)
    channal_num = Data.shape[1]
    signal3d = np.zeros([window_size, channal_num, trialsize * window_num])
    for k in range(len(label)):
        for i in range(window_num):
            signal_label[j] = label[k]
            start = int(pos[k] + i * step + delay)
            signal3d[:, :, j] = Data[start:start + window_size, :]
            j = j + 1


    # data_x = reduceModerate(signal3d, signal_label)
    data_x = normalizeDatax(signal3d)

    # t = np.linspace(1, data_x.shape[0], data_x.shape[0])
    # plt.figure(1)
    # plt.subplot(211)
    # # plt.plot(t[0:400], train_x[0:400, 3, 2])
    # plt.plot(t, data_x[:, 2, 30])
    # plt.subplot(212)
    # # plt.plot(t[0:400], data_x[0:400, 3, 2])
    # plt.plot(t, data_x[:, 3, 30])
    # plt.show()

    return data_x, signal_label


def reduceModerate(signal3d, signal_label):
    # 减静息态均值
    data_x = np.zeros(signal3d.shape)
    samples = signal3d.shape[0]
    sum = np.zeros(signal3d.shape[1])
    for i in range(len(signal_label)):
        if signal_label[i] == 1:
            for j in range(samples):
                sum = sum + signal3d[j, :, i]
            mean = sum / samples
            break
        else:
            continue
    for i in range(len(signal_label)):
        for j in range(samples):
            data_x[j, :, i] = signal3d[j, :, i] - np.transpose(mean)
    return data_x

def normalizeDatax(signal3d):
    # 归一化
    data_x = np.zeros(signal3d.shape)
    channal_num = signal3d.shape[1]
    for i in range(signal3d.shape[2]):
        for j in range(channal_num):
            data_x[:, j, i] = preprocessing.scale(signal3d[:, j, i])
    return data_x


if __name__ == "__main__":
    Data, label = loadov('D:\Myfiles\openvibefiles\MI-CSP-r1\signals\GH\GH-171225-online-1.mat')